### PR TITLE
githubCodeSync: incremental syncing

### LINKED ISSUE #3107: [Caloris2] [Github Code] incremental sync
Daily resync logic

2 possibilities:
- cron like temporal workflow (but meh to add that to the github connector)
- listen for code change webhooks and have a lastCodeSync element per repo to decide whether to start a new sync or not

@fontanierh WDYT? I think I'm in favor of 2?

### PR DESCRIPTION
**Pull request by spolu**
Fixes [#3107](https://github.com/dust-tt/dust/issues/3107)

- Make codeSyncWorkflow signal based (10s debounce)
- Introduce GithubCodeRepository to track last sync times
- Signal the code syncing workflow based on PR webhooks only if repo was not synced for 8h

(the last step can race but since we debounce 10s the workflow, this will result in at most one workflow being started every 8h + no error if there are PR merge events racing through the code path)

Deploy:
- deploy connectors-edge, run init_db
- deploy connectors

### CODE REVIEW

File: connectors/src/api/webhooks/webhook_github.ts [hunk 1]
```diff
@@ -250,6 +254,14 @@ const _webhookGithubAPIHandler = async (
             jsonBody.pull_request.number,
             res
           );
+        } else if (jsonBody.action === "closed") {
```
Comment (philipperolet) [inline connectors/src/api/webhooks/webhook_github.ts:257]:
```suggestion
        } else if (jsonBody.action === "closed" && jsonBody.pull_request.merged) {
```

File: connectors/src/api/webhooks/webhook_github.ts [hunk 2]
```diff
@@ -250,6 +254,14 @@ const _webhookGithubAPIHandler = async (
...
+          return syncCode(
+            enabledConnectors,
+            jsonBody.organization.login,
+            jsonBody.repository.name,
+            jsonBody.repository.id,
+            res
+          );
         } else {
           assertNever(jsonBody.action);
```
Comment (philipperolet) [inline connectors/src/api/webhooks/webhook_github.ts:266]: Unrelated to this PR => we throw for every event that's not opened/edited/closed?

File: connectors/src/api/webhooks/webhook_github.ts (hunk 1 above)
Comment (philipperolet) [reply connectors/src/api/webhooks/webhook_github.ts:257]: avoid syncing on closed without merging PRs

File: connectors/src/connectors/github/temporal/activities.ts [hunk 3]
```diff
@@ -864,16 +865,53 @@ export async function githubCodeSyncActivity({
...
+      where: {
+        connectorId: connector.id,
+        repoId: repoId.toString(),
+      },
+    });
+
     return;
   }
 
+  let githubCodeRepository = await GithubCodeRepository.findOne({
```
Comment (philipperolet) [inline connectors/src/connectors/github/temporal/activities.ts:888]: nit: why not a single upsert?

**Review by philipperolet — CHANGES_REQUESTED** [review]
Looks great ! Partial review, a few little coms
Finishing the review meanwhile

File: connectors/src/api/webhooks/webhook_github.ts (hunk 2 above)
Comment (spolu) [reply connectors/src/api/webhooks/webhook_github.ts:266]: We don't go in that code if it's not an event we want to handle :+1:

File: connectors/src/connectors/github/temporal/activities.ts (hunk 3 above)
Comment (spolu) [reply connectors/src/connectors/github/temporal/activities.ts:888]: I hate upserts :)

File: connectors/src/api/webhooks/webhook_github.ts [hunk 4]
```diff
@@ -381,6 +393,106 @@ async function garbageCollectRepos(
...
+            orgLogin,
+            repoName,
+            repoId,
+          },
+          "githubCodeSync: skipping, GithubCodeRepository not found"
+        );
+        return;
+      }
+
+      const SYNCING_INTERVAL = 1000 * 60 * 60 * 8; // 8 day
```
Comment (philipperolet) [inline connectors/src/api/webhooks/webhook_github.ts]:
```suggestion
      const SYNCING_INTERVAL = 1000 * 60 * 60 * 8; // 8 hours
```

File: connectors/src/connectors/github/temporal/activities.ts (hunk 3 above)
Comment (philipperolet) [reply connectors/src/connectors/github/temporal/activities.ts:888]: Strong point :laughing:

File: connectors/src/api/webhooks/webhook_github.ts (hunk 1 above)
Comment (fontanierh) [reply connectors/src/api/webhooks/webhook_github.ts:257]: If you apply this suggestion I think assertNever will break. Needs to be 2 ifs I think ?

File: connectors/src/connectors/github/temporal/workflows.ts [hunk 5]
```diff
@@ -240,16 +240,15 @@ export async function githubRepoSyncWorkflow(
...
-  await executeChild(githubCodeSyncWorkflow, {
-    workflowId: childWorkflowId,
-    searchAttributes: {
-      connectorId: [parseInt(connectorId)],
-    },
-    args: [dataSourceConfig, githubInstallationId, repoName, repoId, repoLogin],
-    parentClosePolicy: ParentClosePolicy.PARENT_CLOSE_POLICY_TERMINATE,
-    memo: workflowInfo().memo,
+  // Start code syncing activity.
+  await githubCodeSyncActivity({
```
Comment (philipperolet) [inline connectors/src/connectors/github/temporal/workflows.ts:244]: remove `connectorId` from workflow args?

**Review by fontanierh — APPROVED** [review]
LGTM

**Comment by philipperolet** [comment]
Not sure the signalling / debouncing logic is necessary since 2 workflows with the same id can't run at the same time AFAIU? (workflow execution acts as a lock)

File: connectors/src/connectors/github/temporal/workflows.ts (hunk 5 above)
Comment (spolu) [reply connectors/src/connectors/github/temporal/workflows.ts:244]: Not sure I understand?

Comment (spolu) [reply connectors/src/connectors/github/temporal/workflows.ts:244]: Ah I presume you mean that I added it and now can be removed?

Comment (spolu) [reply connectors/src/connectors/github/temporal/workflows.ts:244]: Will do in follow-up PR

**Pull request merged**